### ISSUE TITLE
☂️ FileMaker script steps

### ISSUE THREAD

**Issue by bc-m** [bc-m/fm-xml-export-exploder#28]
https://help.claris.com/en/pro-help/content/script-steps-reference.html

- Control script steps
    - [x] Allow User Abort
    - [ ] Commit Transaction
    - [ ] Configure Local Notification
    - [ ] Configure Machine Learning Model
    - [ ] Configure NFC Reading
    - [ ] Configure Region Monitor Script
    - [x] Else
    - [x] Else If
    - [x] End If
    - [x] End Loop
    - [x] Exit Loop If
    - [X] Exit Script
    - [x] Halt Script
    - [x] If
    - [ ] Install OnTimer Script
    - [x] Loop
    - [ ] Open Transaction
    - [ ] Pause/Resume Script
    - [x] Perform Script
    - [ ] Perform Script On Server
    - [ ] Perform Script On Server with Callback
    - [ ] Revert Transaction
    - [x] Set Error Capture
    - [ ] Set Error Logging
    - [ ] Set Layout Object Animation
    - [x] Set Variable
    - [ ] Trigger Claris Connect Flow
- Navigation script steps
    - [x] Close Popover
    - [x] Enter Browse Mode
    - [x] Enter Find Mode
    - [ ] Enter Preview Mode
    - [x] Go to Field
    - [x] Go to Layout
    - [x] Go to Next Field
    - [x] Go to Object
    - [x] Go to Portal Row
    - [x] Go to Previous Field
    - [x] Go to Record/Request/Page
    - [ ] Go to Related Record
- Editing script steps
    - [ ] Clear
    - [ ] Copy
    - [ ] Cut
    - [ ] Paste
    - [ ] Perform Find/Replace
    - [X] Select All
    - [ ] Set Selection
    - [ ] Undo/Redo
- Fields script steps
    - [ ] Export Field Contents
    - [ ] Insert Audio/Video
    - [ ] Insert Calculated Result
    - [ ] Insert Current Date
    - [ ] Insert Current Time
    - [ ] Insert Current User Name
    - [ ] Insert File
    - [ ] Insert from Device
    - [ ] Insert from Index
    - [ ] Insert from Last Visited
    - [ ] Insert from URL
    - [ ] Insert PDF
    - [ ] Insert Picture
    - [ ] Insert Text
    - [ ] Relookup Field Contents
    - [x] Replace Field Contents
    - [x] Set Field
    - [ ] Set Field By Name
    - [ ] Set Next Serial Value
- Records script steps
    - [x] Commit Records/Requests
    - [ ] Copy All Records/Requests
    - [x] Copy Record/Request
    - [x] Delete All Records
    - [ ] Delete Portal Row
    - [x] Delete Record/Request
    - [x] Duplicate Record/Request
    - [ ] Export Records
    - [ ] Import Records
    - [x] New Record/Request
    - [ ] Open Record/Request
    - [ ] Revert Record/Request
    - [ ] Save Records as Excel
    - [ ] Save Records as PDF
    - [ ] Save Records as Snapshot Link
    - [ ] Truncate Table
- Found Sets script steps
    - [x] Constrain Found Set
    - [x] Extend Found Set
    - [ ] Find Matching Records
    - [x] Modify Last Find
    - [x] Omit Multiple Records
    - [x] Omit Record
    - [x] Perform Find
    - [ ] Perform Quick Find
    - [x] Show All Records
    - [x] Show Omitted Only
    - [ ] Sort Records
    - [ ] Sort Records by Field
    - [x] Unsort Records
- Windows script steps
    - [ ] Adjust Window
    - [ ] Arrange All Windows
    - [x] Close Window
    - [x] Freeze Window
    - [ ] Move/Resize Window
    - [x] New Window
    - [ ] Refresh Window
    - [ ] Scroll Window
    - [ ] Select Window
    - [ ] Set Window Title
    - [ ] Set Zoom Level
    - [ ] Show/Hide Menubar
    - [ ] Show/Hide Text Ruler
    - [ ] Show/Hide Toolbars
    - [ ] View As
- Files script steps
    - [ ] Close Data File
    - [ ] Close File
    - [ ] Convert File
    - [ ] Create Data File
    - [ ] Delete File
    - [ ] Get Data File Position
    - [ ] Get File Exists
    - [ ] Get File Size
    - [x] New File
    - [ ] Open Data File
    - [ ] Open File
    - [ ] Print
    - [ ] Print Setup
    - [ ] Read from Data File
    - [ ] Recover File
    - [ ] Rename File
    - [ ] Save a Copy as
    - [ ] Save a Copy as XML
    - [ ] Set Data File Position
    - [ ] Set Multi-User
    - [x] Set Use System Formats
    - [ ] Write to Data File
- Accounts script steps
    - [ ] Add Account
    - [ ] Change Password
    - [ ] Delete Account
    - [ ] Enable Account
    - [ ] Re-Login
    - [ ] Reset Account Password
    - [ ] Spelling script steps
    - [x] Check Found Set
    - [x] Check Record
    - [ ] Check Selection
    - [x] Correct Word
    - [x] Edit User Dictionary
    - [x] Select Dictionaries
    - [ ] Set Dictionary
    - [x] Spelling Options
    - [ ] Open Menu Item script steps
    - [x] Open Edit Saved Finds
    - [x] Open Favorites
    - [x] Open File Options
    - [x] Open Find/Replace
    - [x] Open Help
    - [x] Open Hosts
    - [x] Open Manage Containers
    - [x] Open Manage Data Sources
    - [x] Open Manage Database
    - [x] Open Manage Layouts
    - [x] Open Manage Themes
    - [x] Open Manage Value Lists
    - [x] Open Preferences
    - [x] Open Script Workspace
    - [x] Open Sharing
    - [x] Open Upload To Host
- Miscellaneous script steps
    - [x] # (Comment)
    - [x] Allow Formatting Bar
    - [ ] AVPlayer Play
    - [ ] AVPlayer Set Options
    - [ ] AVPlayer Set Playback State
    - [x] Beep
    - [ ] Dial Phone
    - [ ] Enable Touch Keyboard
    - [ ] Execute FileMaker Data API
    - [ ] Execute SQL
    - [x] Exit Application
    - [x] Flush Cache to Disk
    - [ ] Get Folder Path
    - [ ] Install Menu Set
    - [ ] Install Plug-In File
    - [ ] Open URL
    - [ ] Perform AppleScript (macOS)
    - [ ] Perform JavaScript in Web Viewer
    - [x] Refresh Object
    - [ ] Refresh Portal
    - [ ] Send DDE Execute (Windows)
    - [ ] Send Event
    - [ ] Send Mail
    - [ ] Set Session Identifier
    - [ ] Set Web Viewer
    - [ ] Show Custom Dialog
    - [ ] Speak (macOS)

**Comment by petrowsky** [comment]
Here is a coverage plan I've worked out with an agent. Please review it and if you want me to move forward give me a thumbs up.

---

# Script Step Coverage Plan

This document captures the strategy for achieving complete script step coverage in `fm-xml-export-exploder`. It is designed as a **session resume point** — read this first when picking up the work in a new session.

The detailed backlog lives in `STUFF_TODO.md`. The per-step implementation process lives in `WORKFLOW.md`. This document covers the *why*, the *order*, and *how the agentic-fm toolchain supports the work*.

Upstream repo: https://github.com/bc-m/fm-xml-export-exploder
Tracking issue: https://github.com/bc-m/fm-xml-export-exploder/issues/28

---

## Current State

Based on the tracking issue as of 2026-03-10:

| Metric | Value |
|--------|-------|
| Steps already implemented | ~140 |
| Steps remaining | ~65 |

The tracking issue checklist and `STUFF_TODO.md` Priority 2 are the authoritative sources for what remains.

---

## Implementation Tiers

Before picking up any step, classify it:

### Tier 1 — Straightforward
No parameters, or only simple boolean/enum flags. The catalog `hrSignature` maps directly to output; the Rust handler is ~10–20 lines. Can usually be implemented without a FileMaker specimen file.

### Tier 2 — Moderate
Requires field refs, calculations, layout refs, or enum-based options. Follows established patterns (see `set_field_data.rs`, `perform_script.rs`). A specimen file is recommended but XML structure can often be inferred from the catalog.

### Tier 3 — Complex / Deferred
AI/ML steps or multi-param data file sequences where the SaXML export format needs real-world validation. Do not implement speculatively — wait for a FileMaker export containing those steps.

---

## Remaining Steps by Batch

Work through `STUFF_TODO.md` Priority 2 in this order. Each batch is a natural PR unit. Update the status marker here as you go.

### Batch 1 — Accounts (4 steps) `[not started]`

Missing: Add Account, Change Password, Re-Login, Reset Account Password

All Tier 1. (Delete Account and Enable Account are already done.) No specimen file needed. Small and self-contained — a good first PR.

### Batch 2 — Small category gaps (3 steps) `[not started]`

Missing: **Editing** — Perform Find/Replace · **Navigation** — Go to Related Record · **Spelling** — Set Dictionary

Three isolated steps across three near-complete categories. Grouping avoids three separate tiny PRs.
- Set Dictionary — Tier 1
- Perform Find/Replace — Tier 2 (calculation + field ref)
- Go to Related Record — Tier 2 (layout + relationship ref)

### Batch 3 — Windows gaps (4 steps) `[not started]`

Missing: Move/Resize Window, Select Window, Set Window Title, Set Zoom Level

Tier 1–2. The other 11 window steps are already done; this closes the category. Patterns from `close_window.rs` and `new_window.rs` apply directly.

### Batch 4 — Control gaps (8 steps) `[not started]`

Missing: Install OnTimer Script, Pause/Resume Script, Perform Script On Server, Perform Script On Server with Callback, Set Layout Object Animation, Configure Region Monitor Script, Configure Local Notification, Configure NFC Reading

Tier 1–3 mixed.
- Pause/Resume Script, Install OnTimer Script — Tier 1–2
- Perform Script On Server / with Callback — Tier 2, follows existing Perform Script pattern
- Configure * steps — Tier 2–3 depending on FM version; check catalog `notes` first
- Trigger Claris Connect Flow has no numeric ID in the catalog — requires special handling, likely deferred

### Batch 5 — Fields gaps (6 steps) `[not started]`

Missing: Insert Audio/Video, Insert File, Insert from URL, Insert PDF, Insert Picture, Relookup Field Contents

All Tier 2. Insert from URL requires the `URL` parameter parser (see `STUFF_TODO.md` Priority 1 — implement that first). Insert File and Insert PDF follow similar patterns. Relookup Field Contents is a simple field reference step.

### Batch 6 — Records gaps (6 steps) `[not started]`

Missing: Delete Portal Row, Export Records, Import Records, Save Records as Excel, Save Records as PDF, Save Records as Snapshot Link

- Delete Portal Row, Save Records as Snapshot Link — Tier 1–2
- Save Records as Excel/PDF — Tier 2, require `UniversalPathList` and `Options` parsers (Priority 1 in `STUFF_TODO.md`)
- Export Records, Import Records — Tier 3 (`Export` and `ImportField` parameter types, Priority 1 in `STUFF_TODO.md`)

Implement the required parameter parsers before the steps that depend on them.

### Batch 7 — Found Sets gap (1 step) `[not started]`

Missing: Sort Records

Tier 2–3. Requires the `SortSpecification` parameter parser (Priority 1 in `STUFF_TODO.md`). Implement that first, then this step is straightforward.

### Batch 8 — Miscellaneous gaps (16 steps) `[not started]`

Missing: AVPlayer Play, AVPlayer Set Options, Execute FileMaker Data API, Execute SQL, Get Folder Path, Install Menu Set, Open URL, Perform AppleScript (macOS), Perform JavaScript in Web Viewer, Refresh Portal, Send DDE Execute (Windows), Send Event, Send Mail, Set Session Identifier, Set Web Viewer, Speak (macOS)

Mixed tiers.
- Get Folder Path, Install Menu Set, Speak, Open URL — Tier 1–2
- Send Mail, Send Event, Execute SQL — Tier 2
- Set Web Viewer, Refresh Portal — Tier 2, require `Object` parser (Priority 1 in `STUFF_TODO.md`)
- Perform JavaScript in Web Viewer — Tier 2, requires `FunctionRef` and `Parameter` parsers (Priority 1)
- Execute FileMaker Data API — Tier 2–3
- Perform AppleScript, Send DDE Execute — platform-specific, Tier 2

### Batch 9 — Files (16 steps) `[not started]`

Missing: Convert File, Create Data File, Delete File, Get Data File Position, Get File Exists, Get File Size, Open Data File, Open File, Print, Print Setup, Read from Data File, Recover File, Save a Copy as, Save a Copy as XML, Set Data File Position, Write to Data File

- Print, Save a Copy as — Tier 2, require `UniversalPathList` and `PageSetup` parsers (Priority 1 in `STUFF_TODO.md`)
- Data file operations (Open/Write/Read/Close/GetPosition/SetPosition) — Tier 3; implement only after obtaining a real FM export containing those steps

### Batch 10 — AI/ML steps `[deferred]`

Configure Machine Learning Model and other AI steps not yet in the enum.

All Tier 3. Defer until a FileMaker 21+ export containing these steps is available.

---

## How agentic-fm Supports This Work

### Step catalog as primary reference

`agent/catalogs/step-catalog-en.json` documents all 205 FileMaker script steps with:
- Numeric step ID (maps directly to `constants.rs` enum value)
- `hrSignature` — the human-readable output format the Rust handler should produce
- `params[]` — full parameter spec: XML element names, types, enum values
- `notes` — behavioral constraints and gotchas

Grep for a step before writing any code:

```bash
grep -A 60 '"name": "Step Name"' /path/to/agentic-fm/agent/catalogs/step-catalog-en.json
```

The `hrSignature` defines what the sanitized output should look like. The `params` array defines which XML child elements to parse. This eliminates most guesswork that would otherwise require manually reading a FileMaker export.

### When a specimen file is still needed

Even with the catalog, a FileMaker specimen export is required when:
1. The step has complex nested parameter structures not fully described in the catalog
2. The step is Tier 3 and the SaXML export format is uncertain
3. You want snapshot test coverage in `tests/xml/`

Follow the manual process in `WORKFLOW.md` ("Prepare a specimen file in FileMaker").

### Existing xml_parsed output as a free specimen

The `agent/xml_parsed/scripts/` folder contains real exported scripts from the Invoice Solution. Before creating a new FM file, check whether any existing scripts already use the target step:

```bash
grep -rl '<Step id="83"' /path/to/agentic-fm/agent/xml_parsed/scripts/
```

If found, that file is a free specimen — no new export needed.

---

## Session Resume Checklist

When starting a new session on this work:

1. Read this file to orient on the current batch and status markers above
2. Check `STUFF_TODO.md` Priority 2 for the next unchecked item in the current batch
3. Grep the step catalog for `hrSignature` and `params`
4. Check `agent/xml_parsed/scripts/` for a free specimen (grep by step ID)
5. If no specimen exists and the step is Tier 2+, create one in FileMaker (see `WORKFLOW.md`)
6. Implement following `WORKFLOW.md` steps 4–9
7. Mark the item done in `STUFF_TODO.md` and on the tracking issue
8. Update the status marker on the batch above

---

## Notes for Upstream Maintainer

This plan was developed using the `agentic-fm` AI-assisted FileMaker development environment, which includes:

- A structured step catalog (`step-catalog-en.json`) covering all 205 current FileMaker script steps with XML parameter specs and HR output signatures
- An AI agent (Claude Code) that can implement Rust handlers given the catalog data and an optional specimen XML file

The implementation approach:
1. The catalog provides the expected HR output format (`hrSignature`) and XML parameter structure for each step
2. For straightforward steps, handlers are written directly from catalog data — no specimen file required
3. For moderate/complex steps, a FileMaker-exported XML specimen validates the handler before commit
4. Each batch above is sized to produce a focused, reviewable PR

Priority 1 parameter parsers (see `STUFF_TODO.md`) are dependencies for several batches and are implemented before the steps that require them — e.g. Batch 7 (Sort Records) requires `SortSpecification`, and Batch 6 (Save Records as PDF) requires `UniversalPathList`.

Batches 1–3 (Accounts, small category gaps, Windows) are entirely self-contained, have no parameter parser dependencies, and would make a clean first set of contributions.
